File: packages/point/python_std/point_std/sql.py
from __future__ import annotations

import json
import os
import re
import sqlite3
from typing import Any


def _open_database() -> sqlite3.Connection | dict[str, str]:
	configured = os.environ.get("POINT_SQL_DATABASE") or os.environ.get("DATABASE_URL") or ":memory:"
	if re.match(r"^postgres(ql)?:", configured, re.IGNORECASE):
		return {
			"message": "std.sql uses SQLite only — declare external postgres driver for PostgreSQL (see docs/site/ecosystem/database-interop.md)",
		}
	path = re.sub(r"^sqlite:", "", configured, flags=re.IGNORECASE)
	try:
		connection = sqlite3.connect(path)
		connection.row_factory = sqlite3.Row
		return connection
	except Exception as error:
		return {"message": str(error)}


def _validate_parameterized_query(sql: str, param_count: int) -> dict[str, str] | None:
	trimmed = sql.strip()
	if len(trimmed) == 0:
		return {"message": "SQL query must not be empty"}
	without_trailing_semicolon = re.sub(r";\s*$", "", trimmed)
	if ";" in without_trailing_semicolon:
		return {"message": "multiple SQL statements are not allowed"}
	placeholders = len(re.findall(r"\?", without_trailing_semicolon))
	if placeholders != param_count:
		return {
			"message": f"parameterized query requires {param_count} ? placeholders, found {placeholders}",
		}
	return None


def _rows_to_json(rows: list[sqlite3.Row]) -> str:
	payload: list[dict[str, Any]] = [dict(row) for row in rows]
	return json.dumps(payload, separators=(",", ":"))


def sqlQueryRaw(sql: str, params: list[str]) -> str | dict[str, str]:
	validation = _validate_parameterized_query(sql, len(params))
	if validation is not None:
		return validation

	db_result = _open_database()
	if isinstance(db_result, dict):
		return db_result

	query_text = re.sub(r";\s*$", "", sql.strip())
	try:
		cursor = db_result.execute(query_text, params)
		if query_text.lstrip().upper().startswith("SELECT"):
			rows = cursor.fetchall()
			return _rows_to_json(rows)
		db_result.commit()
		return "[]"
	except Exception as error:
		return {"message": str(error)}
	finally:
		db_result.close()

File: packages/point/src/std/sql.ts
import { Database } from "bun:sqlite";

type PointStdError = { message: string };

function openDatabase(): Database | PointStdError {
	const configured = process.env.POINT_SQL_DATABASE ?? process.env.DATABASE_URL ?? ":memory:";
	if (/^postgres(ql)?:/i.test(configured)) {
		return {
			message:
				"std.sql uses SQLite only — declare external postgres driver for PostgreSQL (see docs/site/ecosystem/database-interop.md)",
		};
	}
	const path = configured.replace(/^sqlite:/i, "");
	try {
		return new Database(path);
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	}
}

function validateParameterizedQuery(sql: string, paramCount: number): PointStdError | undefined {
	const trimmed = sql.trim();
	if (trimmed.length === 0) {
		return { message: "SQL query must not be empty" };
	}
	const withoutTrailingSemicolon = trimmed.replace(/;\s*$/, "");
	if (withoutTrailingSemicolon.includes(";")) {
		return { message: "multiple SQL statements are not allowed" };
	}
	const placeholders = (withoutTrailingSemicolon.match(/\?/g) ?? []).length;
	if (placeholders !== paramCount) {
		return {
			message: `parameterized query requires ${paramCount} ? placeholders, found ${placeholders}`,
		};
	}
	return undefined;
}

export function sqlQueryRaw(sql: string, params: string[]): string | PointStdError {
	const validation = validateParameterizedQuery(sql, params.length);
	if (validation) {
		return validation;
	}

	const dbResult = openDatabase();
	if ("message" in dbResult) {
		return dbResult;
	}

	const queryText = sql.trim().replace(/;\s*$/, "");
	try {
		const rows = dbResult.query(queryText).all(...params);
		return JSON.stringify(rows);
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	}
}

/** Decode JSON row text from sqlQueryRaw into a runtime array (for typed List<Record> externals). */
export function sqlJsonRowsList(raw: string | PointStdError): unknown[] | PointStdError {
	if (typeof raw === "object" && raw !== null && "message" in raw) {
		return raw;
	}
	try {
		const parsed: unknown = JSON.parse(raw);
		if (!Array.isArray(parsed)) {
			return { message: "SQL rows JSON must be an array" };
		}
		return parsed;
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	}
}
